### PR TITLE
Optional injection

### PR DESCRIPTION
**Pull request by svenjacobs**
Add optional injection via `injectOrNull()`, `injectNowOrNull()` and `getOrNull()`.

### CODE REVIEW

File: core/src/main/kotlin/org/rewedigital/katana/Component.kt [hunk 1]
```diff
@@ -284,18 +309,22 @@ class ComponentContext private constructor(
 ) {
 
     @PublishedApi
-    internal fun <T> injectByKey(key: Key, internal: Boolean = false, arg: Any? = null): T {
+    internal fun <T> findInstance(key: Key, internal: Boolean = false, arg: Any? = null): Instance<T>? {
         val instance = thisComponent.thisComponentInjectByKey<T>(key, internal, arg)
```
Comment (dave08) [inline core/src/main/kotlin/org/rewedigital/katana/Component.kt]: `thisComponentInjectByKey` shouldn't be `thisComponentFindInstance`?

**Review by dave08 — CHANGES_REQUESTED** [review]
Nice :smile:! I'm not sure about my single comment (I don't really have time to look too deeply now), it just seemed to keep consistency with your renaming.

File: core/src/main/kotlin/org/rewedigital/katana/Component.kt (hunk 1 above)
Comment (svenjacobs) [reply core/src/main/kotlin/org/rewedigital/katana/Component.kt]: Yes, you're right. Thanks for the find!

**Comment by svenjacobs** [comment]
> Nice 😄! I'm not sure about my single comment (I don't really have time to look too deeply now), it just seemed to keep consistency with your renaming.

Your remark was correct. I forgot to rename this function 😅

**Review by dave08 — APPROVED** [review]

File: core/src/test/kotlin/org/rewedigital/katana/InjectionTests.kt [hunk 2]
```diff
@@ -266,5 +266,49 @@ object InjectionTests : Spek(
 
                 component.custom<String>(name = "TEST", arg = "World") shouldEqual "Hello World"
             }
+
+            it("inject*OrNull functions should return null for non-declared bindings") {
```
Comment (wengelef) [inline core/src/test/kotlin/org/rewedigital/katana/InjectionTests.kt:270]: is this `*` supposed to be there?

File: core/src/main/kotlin/org/rewedigital/katana/Component.kt [hunk 3]
```diff
@@ -181,6 +197,18 @@ class Component(
...
         context.injectNow<T>(name)
 
+    /**
+     * Injects requested dependency immediately or returns `null` if no binding could be found.
+     *
+     * @see inject
+     * @see injectNow
+     * @throws InstanceCreationException
+     *
+     * TODO: Rename to getOrNull()?
```
Comment (wengelef) [inline core/src/main/kotlin/org/rewedigital/katana/Component.kt:207]: +1 for the api name change to `get`, `getOrNull`, etc.

Comment (svenjacobs) [reply core/src/main/kotlin/org/rewedigital/katana/Component.kt:207]: Thanks for your feedback. The rename will not be part of this PR but I will further think about it.

File: core/src/test/kotlin/org/rewedigital/katana/InjectionTests.kt (hunk 2 above)
Comment (svenjacobs) [reply core/src/test/kotlin/org/rewedigital/katana/InjectionTests.kt:270]: Yes, I mean to test `injectOrNull` and `injectNowOrNull` in this test. Therefore the `*`.

Comment (wengelef) [reply core/src/test/kotlin/org/rewedigital/katana/InjectionTests.kt:270]: got it 👍 although i would separate them then

**Review by wengelef — APPROVED** [review]

**Pull request merged**